### PR TITLE
Fix generating delegates for cases like https://github.com/mono/CppSharp/issues/1144

### PR DESCRIPTION
**Pull request by AleXr64**
```
struct fieldspointers
{
	int(*funcint)(int(*bug)());
};
```
[Result](https://gist.github.com/AleXr64/65b04836e000f4798db48df2cba88faf)
BTW, review requested

### CODE REVIEW

**Comment by ddobrev** [comment]
@AleXr64 thank you for your work but I'm afraid our CI fails. Please ensure all tests pass for you before sending any code.

**Comment by tritao** [comment]
Hey @AleXr64, as @ddobrev said, make sure all existing tests pass, and don't forget to add a new one for the issue that you are fixing.

Also looks like there's some unintended whitespace changes that should be fixed before merging.

**Comment by ddobrev** [comment]
@AleXr64 I'm sorry for this back and forth but I'm afraid you've misunderstood @tritao . Indentation is not just quantity but kind too. You use a tab while we use 4 spaces. You can tell by the diff in this very PR.
In addition, when you feel you're done, please combine all of your commits into one and use your first commit message.

**Comment by AleXr64** [comment]
I do not know why tests are not built on * nix. I guess that due to calling conventions other names are generated. But I do not know how to solve.

**Comment by ddobrev** [comment]
@AleXr64 could you try this instead:

```cs
hasFunctionPtrField.FunctionPtrTakeFunctionPtrField = field => field(); 
Assert.That(hasFunctionPtrField.FunctionPtrTakeFunctionPtrField(() =>  42), Is.EqualTo(42));
```

**Comment by AleXr64** [comment]
Now on OSX build and run. Linux fails with

> 1) Error : /home/travis/build/mono/CppSharp/build/gmake/lib/Release_x64/Common.Tests.CSharp.dll
> System.Runtime.Serialization.SerializationException : The object with ID 2 implements the IObjectReference interface for which all dependencies cannot be resolved. The likely cause is two instances of IObjectReference that have a mutual dependency on each other.
> --SerializationException

**Comment by ddobrev** [comment]
@AleXr64 ignore this one, it's an error from the CI itself. Please just fix the remaining note in my review, combine your commits and add "Fixes <URL_to_your_issue>", this way it will be automatically closed.

File: src/Generator/Passes/DelegatesPass.cs [hunk 1]
```diff
@@ -94,6 +94,8 @@ public override bool VisitFunctionDecl(Function function)
 
         public override bool VisitParameterDecl(Parameter parameter)
         {
+            if(parameter.Namespace?.TranslationUnit?.Module == null && namespaces.Count > 0)
+                parameter.Namespace = namespaces.Peek();
             if (!base.VisitDeclaration(parameter) || parameter.Namespace == null ||
```
Comment (tritao) [inline src/Generator/Passes/DelegatesPass.cs]: Add a new line here.

File: src/Generator/Passes/DelegatesPass.cs [hunk 2]
```diff
@@ -117,11 +119,15 @@ public override bool VisitProperty(Property property)
...
         {
+            if (field.Namespace?.TranslationUnit?.Module != null)
+                namespaces.Push(field.Namespace);
+
             if (!base.VisitFieldDecl(field))
                 return false;
 
             field.QualifiedType = CheckForDelegate(field.QualifiedType,
                 field.Namespace);
+            namespaces.Clear();
```
Comment (tritao) [inline src/Generator/Passes/DelegatesPass.cs]: Same here.

**Comment by tritao** [comment]
I added two small nits as well, just add some extra new lines. Apart from that it looks good, thanks.

**Pull request merged**